src/subsplit_dag.cpp

```diff
@@ -900,3 +914,103 @@ void SubsplitDAG::RemapEdgeIdxs(const SizeVector &edge_reindexer) {
                                      edge_reindexer.at(idx_range.second)};
   }
 }
+
+SubsplitDAG::ListOfNNIs SubsplitDAG::BuildListOfNNIs() {
+  SubsplitDAG::ListOfNNIs list_of_nnis = SubsplitDAG::ListOfNNIs();
+  // Only internal node pairs are viable NNIs.
+  IterateOverRealNodes([&](const SubsplitDAGNode *node) {
+    IterateOverParentAndChildAndLeafwardEdges(
+        node, [&](const size_t parent_id, const bool rotated, const size_t child_id,
+                  const size_t edge_idx) {
+          // Only internal node pairs are viable NNIs.
+          Bitset parent_bitset = GetDAGNode(parent_id)->GetBitset();
+          Bitset child_bitset = GetDAGNode(child_id)->GetBitset();
+          if (!(parent_bitset.SubsplitIsRoot() || child_bitset.SubsplitIsLeaf())) {
+            SafeAddNNIToListOfNNIs(list_of_nnis, parent_bitset, child_bitset, rotated);
+          }
+        });
+  });
+  return list_of_nnis;
+}
+
+void SubsplitDAG::UpdateListOfNNIs(SubsplitDAG::ListOfNNIs &list_of_nnis,
+                                   const Bitset parent_bitset,
+                                   const Bitset child_bitset) {
+  size_t parent_id = GetDAGNodeId(parent_bitset);
+  size_t child_id = GetDAGNodeId(child_bitset);
+  // Every new edge added is a potential new NNI.
+  for (size_t node_id : {parent_id, child_id}) {
+    bool is_parent = (node_id == parent_id);
+    // Get adjacent nodes from both sorted and rotated edges.
+    for (bool is_leafward : {true, false}) {
+      // Get adjacent nodes from both leafward and rootward directions.
+      for (bool is_sorted : {true, false}) {
+        SizeVector adjacent_nodes =
+            GetDAGNode(node_id)->GetLeafwardOrRootward(is_leafward, is_sorted);
```

Comment:
Did you notice that you are passing `is_sorted` to something that's asking for `rotated`?

Reply:
Oh good catch.  Guess it produces the same answer since it adds both sides in the end.  Will fix it.